Define Minetest's direction (#8768)
minetest/minetest: doc/direction.md

# Minetest Direction Document

## 0. Preface

### 0.1. Introduction

This document attempts to set out Minetest's philosophy, aims, and objectives.

Please note that Minetest is developed by a community, and not everyone in the community
holds the same views as to what Minetest should be. Therefore, defining a common direction is

> very difficult.

Reviewer: I think most of this text is obvious and doesn't need to be stated.

> very difficult.

Reviewer 2: Just drop the last message. I think some people should be reminded that not everybody shares their goals.

### 0.2. Definitions

* `Minetest` is the game engine.

> * `Minetest Game` is the default game which comes with the engine.

Reviewer: Yes i think so, as it is still the only bundled game. It needs mentioning as providing some good code examples, although in a poor mod structure.

* The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.

### 0.3. Other documents

> Please read celeron55's document: http://c55.me/blog/?p=1491

Reviewer: This roadmap in the forum is more detailed and a continuation of that blog post https://forum.minetest.net/viewtopic.php?t=9177

Reviewer: Also see celeron55's posts here https://github.com/minetest/minetest/issues/3476 for more roadmap.

Author: I disagree, the article contains more things relevant to direction. The thread is a roadmap, which is distinct to direction. I'll add it anyway

Reviewer: Yes, sorry, it's mostly specific things to implement, but there is also some important 'direction' material in that roadmap if you read through it. So, perhaps this document can do the useful job of extracting the direction stuff from this roadmap?
```
Focus on these kinds of gameplay

    Voxel-based mechanisms that work well in multiplayer environment
        This works well together with the moddability and people have been playing with mesecons, pipes and whatnot for a long time already.
    Defending against computer-controlled enemies
        Defending instead of attacking puts much better use to the aspect of building things... at least until VAEs exist!
        PvP is generally very hard to balance without making it boring
        Many types of this require a world that is more fully in an active state than how Minetest currently works.
            Currently it is assumed that every game wants to maximize world size and player count by taking away from other things, which is not right.
            It should be possible for a mod or game to pick the tradeoff between world size, active entity count, individual entity performance and the resulting maximum player count.
```
```
Some general guidelines

    Networking: The low-level protocol implementation is bloated and not ideal. See extra information in spoiler below. Packet handling, serialization and cross-version compatibility should be made more easily reviewable and more noob-proof.
    Client UI: Keep it generic and customizable so that it works for any game and any server.
    Pull requests: If you test a pull request, always comment on it saying that you did so. The amount of testers is a very important factor in deciding whether to merge or not.
    Maintain fork-friendliness in everything including versioning of things. See http://irc.minetest.ru/minetest-dev/201 ... #i_3654822
    If you want to focus on exploration of the infinite world, you have to put way more effort into emergent generation and active systems in the world in order to actually have infinite things in there. With almost any fixed selection of static things, it's incredibly bland. For the king of this stuff, see Dwarf Fortress.
    Don't mind the lesser visuals and take full advantage of them in how they make creating unimagined things easier. There are enough games in this world that focus on visuals by taking away from everything else, or by making everything else harder to do.
    If you make game content, read this: viewtopic.php?p=139470#p139470
```

Reviewer: .. and also extract the direction-related stuff from #3476

Also see the roadmaps and manifestos of core developers on the forums: https://forum.minetest.net/viewforum.php?f=7

## 1. Mission Statement

Minetest provides both a game engine, which allows the creation of voxel games using Lua, and

> a content distribution platform. Players can connect to servers with their own game play, and

Reviewer 2: The master server is also a service.

easily install content through the Minetest client.

## 2. Goals

* **Stability and performance**, especially on older devices.
* **Keep the engine lightweight**. Support a large amount of gameplay by pushing code and functionality out to Lua,
  rather than having a complex core. Keeping the core as small as possible makes it

>   easier to test and more flexible.

Reviewer 2: I'd rather head into the direction of suggesting to make features as generic as possible to offer many functionalities for less code.

* **Keep the engine as universal as possible**. It's worth comparing everything that goes
  into the engine for how well it suits a gravitation-less space game without ground surface.
  If everything is either disableable or suitable for such, then it gives much more freedom
  for people to experiment and create interesting things, even in ground-based worlds.
* **Make polishing easy**. It shouldn't be made unnecessarily hard to make polished content

>   even while most content isn’t polished.

Reviewer 2: Would you please be so nice to explain me what you mean by this sentence?

Author: it's from c55's roadmap

Reviewer 3: Haha, this sounds like nobody knows what it means, but it has to be carried on for completeness. Maybe there's someone who understands this and who would blame you if you dropped this.

c55 uses "mustn't", "not shouldn't".

I think, what this wants to say is that there's much content (eg. mods code) that uses deprecated stuff, has bad formatting, wasn't updated for a long time and is buggy, but creating new content (eg. a new mod) should be easy to do correct, meaning that you should not need examples to make your own code or that you can easily read what features are deprecated.

Reviewer 4: Sometimes, I picture celeron55 poppin onto the web and thinking... "OMFG, it's _still_ going!?"

The overall "direction" seems to be this way ------>
No, wait.... <-------- that way....
wait... give me a second...

Over the next year, MTG just needs to add some simple mobs (a few peaceful such as animals, and a couple aggressive) whether the pathfinding is perfected or not yet... and then add SmallJokers new bows mod, and players will be happy as a pig in...
...mud

This will keep the hordes at bay for a little while, while everyone figures out wtf to do, to put a solid direction/goals to focus on, whether it be an entirely new game to be packaged or what ever...
Even if it is similar to the current MTG but with a lot of current changes without all the patches & bondo covering the dents, causing bloat.

Reviewer: > Would you please be so nice to explain me what you mean by this sentence?

I think it admits that most MT content isn't very 'polished', but warns against using that fact to justify making polished content excessively difficult to create.
Or maybe, it explains that it is not hypocritical to make polished content possible while most MT content isn't polished.
Makes sense to me.

## 3. Target Games

It's important to limit the scope of the engine to something which is managable.

* Games where players are in 1st or 3rd person.

> * Games with roughly meter cubed cubes.

Reviewer: I agree.

# Minetest Direction Document

---------------------------

## 0. Preface

### 0.1. Introduction

This document attempts to set out Minetest's philosophy, aims, and objectives.

Please note that Minetest is developed by a community, and not everyone in the community
holds the same views as to what Minetest should be.

### 0.2. Definitions

* `Minetest` is the game engine.
* `Minetest Game` is the default game which comes with the engine.

> * The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.

Reviewer 5: I think that "supporting tools" should be defined too. Media server and mod repository are important enough to be mentioned.

### 0.3. Other documents

celeron55 has published his opinions on the development of Minetest in these places:

* What is Minetest? - http://c55.me/blog/?p=1491
* Roadmap - https://forum.minetest.net/viewtopic.php?t=9177
* A clear mission statement for Minetest is missing - https://github.com/minetest/minetest/issues/3476#issuecomment-167399287

Also see the roadmaps and manifestos of core developers on the forums: https://forum.minetest.net/viewforum.php?f=7

---------------------------

> ## 1. Mission Statement

Reviewer 5: I have seen some guys on the forums, who use MT for educational purposes. I think it worth mentioning in mission too.

Reviewer 6: MT For Education isn't a "goal", it's more of a side-effect due to the nature of MT. Facilitating any such "mucking about" is somewhat of the goal of MT, and specific examples of it don't necessarily need to be mentioned.

Reviewer: I agree with benrob0329, use in education is not a primary part of the mission.

Author: Education is an interesting use case but it isn't necessary to mention it

Minetest provides both a game engine, which allows the creation of voxel games using Lua, and

> a content distribution platform. Players can connect to servers with their own gameplay, and

Reviewer 5: > their own gameplay

I think it should be reprased to emphasise the fact that "servers has own gameplay" not that "players has own gameplay".

Reviewer 7: I agree, the wording is ambiguous at present.

easily install content through the Minetest client. The goal is to provide an easy to learn and use,
but still powerful system which allows to easily and rapidly create, extend and modify voxel games.

## 2. Aims

* **Stability and performance**, especially on older devices.
* **Keep the engine as universal as possible**. It's worth comparing everything that goes
  into the engine for how well it suits a gravitation-less space game without ground surface.
  If everything is either disableable or suitable for such, then it gives much more freedom
  for people to experiment and create interesting things, even in ground-based worlds.
* **Keep the engine lightweight**. Support a large amount of gameplay by pushing code and functionality out to Lua,
  rather than having a complex core. Keeping the core as small as possible makes it
  easier to test and more flexible.
* **Make polishing easy**. It shouldn't be made unnecessarily hard to make polished content
  even while most content isn’t polished.

> * **Prefer simplicity**. Try to be simple in both API and features.

Reviewer 5: I think that word "simple" should be elaborated. Simple for who? For example mapgens might be simple for paramat, but very complex for me. Does mapgen algorithm contradicts to "simplicity" aim?
Software is complex by nature. Implementation of interesting feature might require sophisticated technical details.

I think it is better to write something like "API should be documented. Each function should be provided with examples on wiki. Each development feature should be added for one purpose so modders could combine them to make something complex."

Reviewer 7: > For example mapgens might be simple for paramat, but very complex for humans

FTFY :)

* **Self-contained**. Avoids relying on third party behaviour. For example, there is no
  central account system and the serverlist isn't actually needed to play online.
* **Multi-platform**. Essential features should work on all platforms.

## 3. Scope of Game Engine

The engine is designed for games which fit these rough limitations:

> * Players are in 1st or 3rd person.

Reviewer 5: What about isometric view? What about RTS-style kind of gameplay?

> * Roughly meter sized cubes.

Reviewer 5: Does it even matter?
Cubes which feel like 0.5 meter size are prohibited to implement?
Large player model, which will have dimensions like 2x2x3 is prohibited?

Reviewer 2: No, cubes as in `drawtype = "normal"` are engine-defined. One distance in the grid is one meter per definition.

Reviewer 5: Still, some might want to implement "huge world". In such world, each cube will be equal to 1 inch and nothing will actually change from engine point of view.

# Minetest Direction Document

---------------------------

## 0. Preface

### 0.1. Introduction

This document attempts to set out Minetest's philosophy, aims, and objectives.

Please note that Minetest is developed by a community, and not everyone in the community

> holds the same views as to what Minetest should be.

Reviewer: This paragraph states the obvious so i think it is unnecessary. It also seems awkward and over-apologetic about these obvious and unavoidable differences.

### 0.2. Definitions

* `Minetest` is the game engine.
* `Minetest Game` is the default game which comes with the engine, and is not covered by this document.
* The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.

> 

Reviewer 6: ```suggestion
* `Content` refers to any installable add-ons, such as Mods, Games, and Texture Packs.

```

### 0.3. Other documents

In order to understand this document, it's important to get context.

celeron55 has published his opinions on the development of Minetest in these places:

  * What is Minetest? - http://c55.me/blog/?p=1491
  * Roadmap - https://forum.minetest.net/viewtopic.php?t=9177
  * A clear mission statement for Minetest is missing - https://github.com/minetest/minetest/issues/3476#issuecomment-167399287

Also see the roadmaps and manifestos of core developers on the forums: https://forum.minetest.net/viewforum.php?f=7

---------------------------

## 1. Mission Statement

Minetest provides both a game engine, which allows the creation of voxel games using Lua, and
a content distribution platform. Players can connect to servers which provide their own gameplay,
and easily install content through the Minetest client. The goal is to provide an easy to learn and use,
but still powerful system which allows to easily and rapidly create, extend and modify voxel games.

## 2. Aims

1. **Stability and performance**,

> 		especially on older devices.

Reviewer 6: ```suggestion
		Mobile and low-end devices should not be excluded from essential graphical and gameplay elements.
```

> 		especially on older devices.

Reviewer 8: not older but low end yes.
I'm running my minetest server on rock 64 (which is currently the most powerful balanced with ram & cpu low end device server we can host at home), and minetest server should run fast on it.
For the client android is the good low end device target.
Note: currently running the most interesting mods on the rock 64 (and i think the RPI 4 too) cause minetest server to lag with 1 or 2 servers. I mean with item_drop which is just essential in survival farmining mod, and mob frameworks too (the get_obj_in_radius PR has improved the perf but we are very very far from disabling both mods)

2. **Keep the engine as universal as possible**.
		It's worth comparing everything that goes into the engine for how well
		it suits a gravitation-less space game without ground surface.
		If everything is either disableable or suitable for such, then it gives
		much more freedom for people to experiment and create interesting things,
		even in ground-based worlds.
3. **Keep the engine lightweight**.
		Support a large amount of gameplay by pushing code and functionality out
		to Lua, rather than having a complex core. Keeping the core as small as

> 		possible makes it easier to test and more flexible.

Reviewer 9: I'd like to have some balance mentioned here
since performance also matters, moving some features into the core <b>is</b> a good idea (even if lua could technically do it on its own)

Reviewer 8: i don't agree with keeping engine lightweight, it's a wrong direction regarding the performance. I'd rather want to see generic more than lightweight. By generic i mean no explicit object, like firespread, chests or something like this.
But higher level Lua calls are always needed for the performance, and we may need also some frameworks to help minetest to become a game. We are not a game engine, we are mainly tied to voxel games, and we should offer features needed to do a finished an performant game for end users.

Reviewer 8: The core complexity is currently due to a big code layering due to historic and non refactored code from the 0.3 era when minetest moved from a single game to a multi player game. We worked hard to improve it, but we have some layers which are currently crazy to maintain. 
For me, mapgen is one of the good candidates for a refactor (the mapblock generation process), but also the network layer, and the ServerEnvironment.

4. **Prefer simplicity**.

> 		Try to be simple in both API and implementing features.

Reviewer 6: ```suggestion
3. **Keep The Core Simple**.
		Prefer simpler solutions and APIs; keeping the core as small as possible
		makes it easier to test and maintain. Pushing unneeded features to Lua
		allows mods and games to better implement custom behavior.
4. **Flexible APIs**.
		Support a wide variety of gameplay paradigms by making generic APIs
		that are easily configured or scripted in Lua.
```

Reviewer 8: simple API != performance. We should have `coherent API` more than simple API. low level API means we will deny responsilibility and let modders do the same code again and again.
When some code it too repeated and can be integrated in a fashioned and performant way to the core, we should do it.

5. **Self-contained**.

> 		Avoids relying on third party behaviour. For example, there is no

Reviewer 6: ```suggestion
		Avoid relying on third party behavior. For example, there is no
```

> 		central account system and the serverlist isn't actually needed to play online.

Reviewer 6: ```suggestion
		central account system and the serverlist isn't needed to play online.
```

Reviewer 8: i think we can say there will never be a central account system. We don't have the resources to manage that (in terms of time & money)

6. **Multi-platform**. Essential features should work on all platforms.

> 7. **Gameplay over graphics**.

Reviewer 5: You could write something about "encouraging imagination" by graphics in MTG and default game.
Simplicity of MTG might be pretended to reduce entry threshold for new modders.
Detailisation of other games does not have to be primitive, but should be consistent.

> 7. **Gameplay over graphics**.

Reviewer 6: ```suggestion
7. **Flexible, Low-Tech Graphics**.
		Because of the need to support mobile and low-end devices, graphics should
		be suitably low-tech to maintain performance and consistency, however despite
		this the APIs given should be flexible enough that content can make the most
		of the engine's capabilities.
```

Reviewer 8: i think we can say also that we can have nice graphics but it should be toggled by some options and not enabled by default for supporting various devices

> 8. **Maintain fork-friendliness**.

Reviewer 5: Is this something about "if team denies your PR and you made your own fork, then one day our team might merge your work to master"?

Reviewer 6: This line seems useless without explanation or context.
```suggestion
8. **Don't Be Hostile Towards Project Forks**.
		Attitude towards divergent code-bases and communities should remain friendly,
		and changes should not be made to spite these.
```

Reviewer: It is better this line is stated in the positive "fork-friendliness" instead of the negative "do not be hostile", "... to spite these". The suggestion above is unpleasant and dramatic.
Besides that, this may not be what "fork-friendliness" is about.

Reviewer 6: You're probably right, a positive statement will sound better than a negative one. Perhaps Ruben can give some insight into what he meant.

Reviewer 8: we must keep positive statement.
Also we should encourage backporting fixes & features as far it's possible

## 3. Scope of Game Engine

The engine is designed for games which fit these rough limitations:

* The world consists of a 3d grid of discrete positions, the distance between each position
  is called a unit length.
* Each position is filled by exactly one node. Positions that appear to be empty are actually filled with airlike nodes.
  Nodes may exceed their bounds, but still belong to one position.
* The visible size of a unit length may change according to the game's requirements, but the engine does
  not aim to promise good performance for visible sizes far deviating the default.

> * Games will tend to define the unit length distance a meaning, for example, Minetest Game defines.

Reviewer 9: Minetest Game defines?

> * Games will tend to define the unit length distance a meaning, for example, Minetest Game defines.

Reviewer 6: ```suggestion
* Games will tend to define the unit length distance a meaning, for example, Minetest Game defines one node as one meter.
```

## 4. Future Work

* **Improving flexibility**.
		To fulfill aim 2, development will seek to push functionality to Lua and increase the flexibility

> 		and potential.

Reviewer 6: ```suggestion
		To fulfill aim 2, development will seek to de-hardcode and increase the flexibility of the Lua API.
```

* **Improved graphics and rendering performance**
		Development will seek to improve the performance of rendering, to increase FPS on both old and new

> 		devices. New graphical features may be added if they do not jeapordize performance when not used.

Reviewer 9: If this were true, shader would have been added.
this needs less absolute terms

Reviewer 3: Also that word should probably be "je**o**p**a**rdize".

Reviewer 6: sfan5: Not entirely true. Shaders are fundamental to any non-ancient rendering pipeline, and can actually help performance when used correctly. MT does not use them correctly unfortunately.

* **Improve code structure and maintainability**

> 		Development will seek to improve the maintainability of code through refactoring.

Reviewer 5: > Development will seek to improve the maintainability of code through refactoring.

Before you'll start refactoring, you should write TESTS. Auto tests are preferrable, but keeping suite of manual test-scenarios might be good for a start.

Author: most of the code doesn't lend well to automated tests

Reviewer 8: currently many parts of the code are too heavy to support tests, but refactored one should provide tests. We don't enforce them but we encourage when it's possible and we enforced when it's necessary

> 		Development will seek to improve the maintainability of code through refactoring.

Reviewer 6: ```suggestion
		Development will seek to improve the maintainability of code through refactoring and the addition of Unit Tests.
```

a content distribution platform. Players can connect to servers which provide their own gameplay,
and easily install content through the Minetest client. The goal is to provide an easy to learn and use,
but still powerful system which allows to easily and rapidly create, extend and modify voxel games.

## 2. Aims

What should changes to the engine take into mind?

1. **Stability and performance**.
		Mobile and low-end devices should not be excluded from essential graphical
		and gameplay elements.
2. **Keep the engine as universal as possible**.
		It's worth comparing everything that goes into the engine for how well
		it suits a gravitation-less space game without ground surface.
		If everything is either disableable or suitable for such, then it gives
		much more freedom for people to experiment and create interesting things,
		even in ground-based worlds.
3. **Keep the engine lightweight**.
		Support a large amount of gameplay by pushing code and functionality out
		to Lua, where possible, rather than having a complex core.
		The exception to this is when APIs are needed to increase performance.

> 		Keeping the core as small as possible makes it easier to test and more flexible.

Reviewer 6: This section feels like it belongs in the sections preceding and following it. Keeping features in Lua sounds like a flexibility problem and not a performance one, and thus belongs in the sections referencing a gravitation-less space game. Similarly, stating that it's easier to maintain sounds like it belongs in the "Prefer simplicity" section as that's about code simplicity.

Basically, what is this section saying that isn't already, or couldn't just be stated in the other two sections?
